src/System/Process/Typed.hs

```diff
@@ -750,27 +754,73 @@ stopProcess = liftIO . pCleanup
 -- | Uses the bracket pattern to call 'startProcess' and ensures that
 -- 'stopProcess' is called.
 --
--- In version 0.2.0.0, this function was monomorphized to @IO@ to
--- avoid a dependency on the exceptions package.
+-- This function is usually /not/ what you want. You're likely better
+-- off using 'withProcessWait'. See
+-- <https://github.com/fpco/typed-process/issues/25>.
+--
+-- @since 0.2.5.0
+withProcessTerm
+  :: ProcessConfig stdin stdout stderr
+  -> (Process stdin stdout stderr -> IO a)
+  -> IO a
+withProcessTerm config = bracket (startProcess config) stopProcess
+
+-- | Uses the bracket pattern to call 'startProcess'. Unlike
+-- 'withProcessTerm', this function will wait for the child process to
+-- exit, and only kill it with 'stopProcess' in the event that the
+-- inner function throws an exception.
+--
+-- @since 0.2.5.0
+withProcessWait
+  :: ProcessConfig stdin stdout stderr
+  -> (Process stdin stdout stderr -> IO a)
+  -> IO a
+withProcessWait config f =
+  bracket
+    (startProcess config)
+    stopProcess
+    (\p -> f p <* waitExitCode p)
+
+-- | Deprecated synonym for 'withProcessTerm'.
 --
 -- @since 0.1.0.0
 withProcess :: ProcessConfig stdin stdout stderr
             -> (Process stdin stdout stderr -> IO a)
             -> IO a
-withProcess config = bracket (startProcess config) stopProcess
+withProcess = withProcessTerm
+{-# DEPRECATED withProcess "Please consider using withProcessWait, or instead use withProcessTerm" #-}
 
--- | Same as 'withProcess', but also calls 'checkExitCode'
+-- | Same as 'withProcessTerm', but also calls 'checkExitCode'
 --
--- In version 0.2.0.0, this function was monomorphized to @IO@ to
--- avoid a dependency on the exceptions package.
+-- @since 0.2.5.0
+withProcessTerm_
+  :: ProcessConfig stdin stdout stderr
+  -> (Process stdin stdout stderr -> IO a)
+  -> IO a
+withProcessTerm_ config = bracket
+    (startProcess config)
+    (\p -> stopProcess p `finally` checkExitCode p)
+
+-- | Same as 'withProcessWait', but also calls 'checkExitCode'
+--
+-- @since 0.2.5.0
+withProcessWait_
+  :: ProcessConfig stdin stdout stderr
+  -> (Process stdin stdout stderr -> IO a)
+  -> IO a
+withProcessWait_ config f = bracket
+    (startProcess config)
+    stopProcess
+    (\p -> f p <* checkExitCode p)
```

Comment:
Any reason why on this case (and in `withProcessWait`), you have chosen `<*` as compared to `finally` ?

Reply:
Yes. If there was an exception in the inner block, we want that exception to propagate. Checking the exit code of the process is irrelevant when we're already about to throw a runtime exception.